Mathlib/Combinatorics/SimpleGraph/Operations.lean

```diff
@@ -55,4 +56,86 @@ lemma adj_replaceVertex_iff_of_ne_right {w : V} (hw : w ≠ t) :
+  ext e
+  refine' e.inductionOn _
+  simp only [Set.mem_toFinset, mem_edgeSet, mem_union, mem_sdiff, mem_incidenceFinset,
+    mk'_mem_incidenceSet_iff]
+  intros; split_ifs; exacts [by simp_all, by aesop, by rw [adj_comm]; aesop, by aesop]
+
+theorem edgeFinset_replaceVertex_of_adj (ha : G.Adj s t) :
+    (G.replaceVertex s t).edgeFinset = (G.edgeFinset \ G.incidenceFinset t ∪
+      (G.neighborFinset s).image fun v ↦ s(v, t)) \ {s(t, t)} := by
+  ext e
+  refine' e.inductionOn _
+  simp only [Set.mem_toFinset, mem_edgeSet, mem_union, mem_sdiff, mem_incidenceFinset,
+    mk'_mem_incidenceSet_iff]
+  intros; split_ifs; exacts [by simp_all, by aesop, by rw [adj_comm]; aesop, by aesop]
+
+lemma disjoint_sdiff_neighborFinset_image : Disjoint (G.edgeFinset \ G.incidenceFinset t)
+    ((G.neighborFinset s).image fun v ↦ s(v, t)) := by
+  rw [disjoint_iff_ne]
+  intro e he
+  have : t ∉ e := by
+    rw [mem_sdiff, mem_incidenceFinset] at he
+    obtain ⟨_, h⟩ := he
+    contrapose! h
+    simp_all [incidenceSet]
+  aesop
+
+theorem card_edgeFinset_replaceVertex_of_not_adj (hn : ¬G.Adj s t) :
+    (G.replaceVertex s t).edgeFinset.card = G.edgeFinset.card + G.degree s - G.degree t := by
+  have inc : G.incidenceFinset t ⊆ G.edgeFinset := by simp [incidenceFinset, incidenceSet_subset]
+  rw [G.edgeFinset_replaceVertex_of_not_adj hn,
+    card_disjoint_union G.disjoint_sdiff_neighborFinset_image, card_sdiff inc,
+    tsub_add_eq_add_tsub <| card_le_card inc, card_incidenceFinset_eq_degree]
+  congr 2
+  rw [card_image_of_injective, card_neighborFinset_eq_degree]
+  unfold Function.Injective
+  aesop
+
+theorem card_edgeFinset_replaceVertex_of_adj (ha : G.Adj s t) :
+    (G.replaceVertex s t).edgeFinset.card = G.edgeFinset.card + G.degree s - G.degree t - 1 := by
+  have inc : G.incidenceFinset t ⊆ G.edgeFinset := by simp [incidenceFinset, incidenceSet_subset]
+  rw [G.edgeFinset_replaceVertex_of_adj ha, card_sdiff (by simp [ha]),
+    card_disjoint_union G.disjoint_sdiff_neighborFinset_image, card_sdiff inc,
+    tsub_add_eq_add_tsub <| card_le_card inc, card_incidenceFinset_eq_degree]
+  congr 2
+  rw [card_image_of_injective, card_neighborFinset_eq_degree]
+  unfold Function.Injective
+  aesop
+
 end ReplaceVertex
+
+section AddEdge
+
+/-- Given a vertex pair, add the corresponding edge to the graph's edge set if not present. -/
+abbrev addEdge : SimpleGraph V where
+  Adj v w := G.Adj v w ∨ s ≠ t ∧ (s = v ∧ t = w ∨ s = w ∧ t = v)
+  symm v w := by simp_rw [adj_comm]; (conv_lhs => arg 2; arg 2; rw [or_comm]); exact id
+
+@[simp]
+lemma addEdge_self : G.addEdge s s = G := by ext; simp
+
+lemma addEdge_adj (h : G.Adj s t) : G.addEdge s t = G := by
+  ext
+  simp only [ne_eq, G.ne_of_adj h, not_false_eq_true, true_and, or_iff_left_iff_imp]
+  rintro (_ | _) <;> simp_all [adj_comm]
+
+variable [Fintype V] {s t} [DecidableRel G.Adj]
+
+theorem edgeFinset_addEdge (hn : ¬G.Adj s t) (h : s ≠ t) :
+    (G.addEdge s t).edgeFinset = G.edgeFinset ∪ {s(s, t)} := by
```

Comment:
Can you use `Finset.cons` here, to avoid the `DecidableRel`?

Reply:
All done!